tests/unit/transaction_record_test.py

```
@@ -389,3 +413,95 @@ def test_proto_conversion_with_call_result():
+    dup = TransactionRecord(transaction_id=transaction_id)
+    record.duplicates = [dup, dup]
+
+    assert "duplicates_count=2" in repr(record), "duplicates_count should reflect list length"
+    
+def test_from_proto_raises_when_no_transaction_id_available():
+    """Verify error is raised when neither transaction_id param nor proto.transactionID is present."""
+    proto = transaction_record_pb2.TransactionRecord()
+    
+    assert not proto.HasField("transactionID")
+    
+    with pytest.raises(ValueError, match="transaction_id is required when proto.transactionID is not present"):
```

Ruff (0.14.14)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)